src/aleph/http/multipart.clj

```diff
@@ -127,37 +129,68 @@
       (.flip buf)
       (bs/to-byte-array buf))))
 
-(defn encode-request [^DefaultHttpRequest req parts]
+(defn- ensure-charset [charset]
+  (cond
+    (nil? charset)
+    HttpConstants/DEFAULT_CHARSET
+
+    (string? charset)
+    (Charset/forName charset)
+
+    (instance? Charset charset)
+    charset))
+
+(defn- encode-file-upload
+  "Encode the `content` as a `FileUpload`."
+  [^HttpPostRequestEncoder encoder {:keys [part-name name content mime-type charset]}]
+  (let [filename  (.getName ^File content)
+        name      (or name filename)
+        mime-type (or mime-type
+                      (URLConnection/guessContentTypeFromName filename))
+        content-type (mime-type-descriptor mime-type charset)]
+    (.addBodyFileUpload encoder
+                        (or part-name name)
+                        ;; Netty's multipart encoder ignores empty strings here
+                        (or filename "")
+                        content
+                        content-type
+                        false)))
+
+(defn- encode-memory-file-upload
+  "Encode the `content` as a `MemoryFileUpload`."
+  [^HttpPostRequestEncoder encoder {:keys [part-name name content mime-type charset]}]
+  (let [content-type (mime-type-descriptor mime-type charset)
+        charset      (ensure-charset charset)
+        filename     (str (UUID/randomUUID))
```

Comment:
Just saw this. Why is `filename` a random string? Why not expose it as a parameter to achieve parity with Apache's `ByteArrayBody`?

Reply:
Arnaud will know better, but my guess is it's because it's for a MemoryFileUpload, which mimics a file in memory, and as such, the "filename" is just for compatibility with the FileUpload interface, since there's not really a file.

OTOH, I can imagine a real filename being useful. Want to submit a PR?

Reply:
> Arnaud will know better, but my guess is it's because it's for a MemoryFileUpload, which mimics a file in memory, and as such, the "filename" is just for compatibility with the FileUpload interface, since there's not really a file.

Exactly, that was the intent behind it.

Reply:
Created a PR: https://github.com/clj-commons/aleph/pull/668